File: apps/ios/Sources/SwabCore/Onboarding/OnboardingState.swift
/// ONB-08 — resumable onboarding. The current step is persisted locally
/// (plain: a step name is not classification data). Killing the app mid-flow
/// resumes at the same step via the root gate.
///
/// The step stays `.phone` until OTP verification succeeds — the pending
/// phone hash is memory-only (`PendingSignup`), so a restart during OTP
/// re-asks the number.
import Foundation

public enum OnboardingStep: String, CaseIterable, Sendable {
    case welcome
    case phone
    case contacts
    case calibrate
    case done
    case complete
}

public actor OnboardingStateStore {
    private static let stepKey = "onboarding.step.v1"

    private let kv: KeyValueStore
    private var cached: OnboardingStep?

    public init(kv: KeyValueStore) {
        self.kv = kv
    }

    public func getStep() async -> OnboardingStep {
        if let cached {
            return cached
        }
        let raw = await kv.get(Self.stepKey)
        let step = raw.flatMap { OnboardingStep(rawValue: $0) } ?? .welcome
        cached = step
        return step
    }

    public func setStep(_ step: OnboardingStep) async {
        cached = step
        await kv.set(Self.stepKey, value: step.rawValue)
    }

    /// Test seam: drops in-memory state, simulating a process restart.
    public func resetForTests() {
        cached = nil
    }
}

public func route(for step: OnboardingStep) -> String {
    switch step {
    case .welcome: return "/onboarding/welcome"
    case .phone: return "/onboarding/phone"
    case .contacts: return "/onboarding/contacts"
    case .calibrate: return "/onboarding/calibrate"
    case .done: return "/onboarding/done"
    case .complete: return "/"
    }
}

File: apps/android/app/src/main/kotlin/com/swab/android/onboarding/OnboardingState.kt
package com.swab.android.onboarding

import com.swab.android.storage.KeyValueStore

/**
 * ONB-08 — resumable onboarding. The current step is persisted locally
 * (plain: a step name is not classification data). Killing the app mid-flow
 * resumes at the same step. Port of apps/mobile/src/onboarding/state.ts.
 *
 * Note: the step stays [OnboardingStep.PHONE] until OTP verification
 * succeeds — the pending phone hash is memory-only (see PendingSignup), so a
 * restart during OTP re-asks the number.
 */
enum class OnboardingStep(val wireName: String) {
    WELCOME("welcome"),
    PHONE("phone"),
    CONTACTS("contacts"),
    CALIBRATE("calibrate"),
    DONE("done"),
    COMPLETE("complete"),
    ;

    companion object {
        fun fromWireName(value: String?): OnboardingStep =
            entries.firstOrNull { it.wireName == value } ?: WELCOME
    }
}

class OnboardingStateStore(private val kv: KeyValueStore) {
    companion object {
        private const val STEP_KEY = "onboarding.step.v1"
    }

    private var cached: OnboardingStep? = null

    suspend fun getStep(): OnboardingStep {
        val existing = cached
        if (existing != null) return existing
        val raw = kv.get(STEP_KEY)
        val step = OnboardingStep.fromWireName(raw)
        cached = step
        return step
    }

    suspend fun setStep(step: OnboardingStep) {
        cached = step
        kv.set(STEP_KEY, step.wireName)
    }
}
